cpp/src/join/joining.cu

```diff
@@ -399,83 +393,121 @@ gdf_error construct_join_output_df(
       CHECK_STREAM(0);
       gdf_error update_err = nvcategory_gather_table(join_source_table,join_destination_table);
       CHECK_STREAM(0);
-      GDF_REQUIRE(update_err == GDF_SUCCESS,update_err);
+      CUDF_EXPECTS(update_err == GDF_SUCCESS, "nvcategory_gather_table throwing a GDF error");
     }
-
+     
     POP_RANGE();
-    return GDF_SUCCESS;
+    return std::pair<cudf::table, cudf::table>(result_left, result_right);
 }
 
 template <JoinType join_type, typename index_type>
-gdf_error join_call_compute_df(
-                         gdf_column **left_cols, 
-                         int num_left_cols,
-                         int left_join_cols[],
-                         gdf_column **right_cols,
-                         int num_right_cols,
-                         int right_join_cols[],
-                         int num_cols_to_join,
-                         int result_num_cols,
-                         gdf_column **result_cols,
-                         gdf_column * left_indices,
-                         gdf_column * right_indices,
+std::pair<cudf::table, cudf::table> join_call_compute_df(
+                         cudf::table const& left_cols, 
+                         cudf::table const& right_cols,
+                         cudf::table const& join_cols,
+                         cudf::table const& merging_join_cols,
+                         cudf::table const& out_indices, 
                          gdf_context *join_context) {
-  GDF_REQUIRE(nullptr != left_cols, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(nullptr != right_cols, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(0 != num_cols_to_join, GDF_SUCCESS);
-  GDF_REQUIRE(nullptr != left_join_cols, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(nullptr != right_join_cols, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(nullptr != join_context, GDF_INVALID_API_CALL);
 
-  for(int column_index = 0; column_index  < num_left_cols; column_index++){
-    GDF_REQUIRE(left_cols[column_index]->dtype != GDF_invalid,GDF_UNSUPPORTED_DTYPE);
+  int num_left_cols = left_cols.num_columns();
+  int num_right_cols = right_cols.num_columns();
+  int num_cols_to_join = (join_cols.num_columns() > 0 ? join_cols.get_column(0)->size: 0);
+  int num_merging_join_cols = (merging_join_cols.num_columns() > 0 ? merging_join_cols.get_column(0)->size: 0);
+
+  std::vector<int> left_join_cols (num_cols_to_join);
+  std::vector<int> right_join_cols (num_cols_to_join);
+  std::vector<int> l_common_name_join_ind (num_merging_join_cols);
+  std::vector<int> r_common_name_join_ind (num_merging_join_cols);
+
+  if (num_cols_to_join > 0)
+  {
+      CUDA_TRY (cudaMemcpy((void *)left_join_cols.data(), (void *)join_cols.get_column(0)->data, 
```

Comment:
Here again copying something that might have a larger size than the allocated size of the target vector.